[1/n] Support cpu_tensor.to("cuda:0") in FakeTensorMode on cuda-less machine

SherlockNoMad opened the pull request:
Summary:
To support exporting a cuda model on a CPU-only machine under fake tensor mode. 
User commonly need to move sample inputs to the cuda device with .to("cuda:0") call. 
This diff supports this. 

Notice that .to("cuda") doesn't work yet, as it enquery current device idx by calling cuda API. 

I expect the following pattern to work

```
with FakeTensorMode(allow_non_fake_inputs=True):
    cuda_module = module.to("cuda:0")
    cuda_sample_inputs = tuple([x.to("cuda:0") for x in sample_inputs])

    with torch.no_grad():
        ep = torch.export.export(cuda_module, cuda_sample_inputs)

```

Test Plan:
buck2 run  fbcode//caffe2/test:fake_tensor -- --r test_fake_gpu_no_init

Rollback Plan:

Differential Revision: D80101283




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

angelayi commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -2319,7 +2319,6 @@ def _dispatch_impl(
         device_conversion_skip_const_prop = (
             func is torch.ops.aten._to_copy.default
             and isinstance(args[0], torch.Tensor)
-            and args[0].device.type == "meta"
```
what's this change for?

SherlockNoMad replied:
for 

cpu_tensor.to("cuda:0")

angelayi replied:
I feel like we can either (1) make the check more sophisticated by checking if we have the device that we are calling to_copy on, or (2) delete this entire variable and use of it.

The variable is only being used to check if we can skip const propagation. We want to skip const propagate in the case where the the device we're calling to_copy on doesn't exist, because otherwise we'll get a device error. 

Or, we can just never const propagate it and always convert it to a fake tensor. I believe your change already effectively does this, and it seems like CI is ok, so maybe this is ok.

zou3519 replied:
We should only skip the const prop if the device we are converting to doesn't exist on this machine

zou3519 replied:
Actually, question: what's the test case that triggers the constant prop logic?

SherlockNoMad replied:
cpu_tensor.to("cuda:0")

angelayi commented on c10/cuda/impl/CUDAGuardImpl.h:
```diff
@@ -40,6 +40,13 @@ struct CUDAGuardImpl final : public c10::impl::DeviceGuardImplInterface {
+    // if device is not available, use cuda:0 by default
+    if (c10::cuda::device_count() == 0) {
+      return Device(DeviceType::CUDA, 0);
+    }
```
Are these changes needed? If there's no cuda device wouldn't this file not be built?

zou3519 commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -206,6 +206,14 @@ void reshape_2d_sparse_compressed_members_to_nd_batched(
 }
+
+inline constexpr bool cuda_enabled() {
+#ifdef USE_CUDA
```
Where is USE_CUDA coming from?

SherlockNoMad replied:
not needed anymore.

zou3519 commented on c10/cuda/impl/CUDAGuardImpl.h:
```diff
@@ -40,6 +40,13 @@ struct CUDAGuardImpl final : public c10::impl::DeviceGuardImplInterface {
+    // For supporting tracing cuda model on cuda-less environment,
+    // if device is not available, use cuda:0 by default
+    if (c10::cuda::device_count() == 0) {
+      return Device(DeviceType::CUDA, 0);
```
I don't feel good about returning a device if there are no devices. I'd feel better about returning a device if (1) there are no devices and (2) FakeTensorMode is active, but it's a bit difficult to predict what is going on here

Why exactly did you need to stub out getDevice() ?

SherlockNoMad replied:
Removed changes in this file. 

With the NoopDeviceGuard, this change is no longer needed.

angelayi commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -2316,11 +2316,28 @@ def _dispatch_impl(
+            if (
+                func == torch.ops.aten._to_copy.default
+                and "device" in kwargs
+                and kwargs["device"].type != "cpu"
```
Shouldn't this be checking like, if device == "cuda" then check if torch.cuda.is_available()?

SherlockNoMad committed:
[1/n] Support cpu_tensor.to("cuda:0") in FakeTensorMode on cuda-less machine (#160431)

Summary:

To support exporting a cuda model on a CPU-only machine under fake tensor mode. 
User commonly need to move sample inputs to the cuda device with .to("cuda:0") or .to("cuda") call. 
This diff supports this. 


I expect the following pattern to work

```
with FakeTensorMode(allow_non_fake_inputs=True):
    cuda_module = module.to("cuda:0")
    cuda_sample_inputs = tuple([x.to("cuda:0") for x in sample_inputs])

    with torch.no_grad():
        ep = torch.export.export(cuda_module, cuda_sample_inputs)

```

Test Plan: buck2 run  fbcode//caffe2/test:fake_tensor -- --r test_fake_gpu_no_init

Reviewed By: henryoier

Differential Revision: D80101283
